### ISSUE TITLE
Accessibility issues with site design and implementation

### ISSUE THREAD

**Issue by thibaudcolas** [django/djangoproject.com#1381]
During DjangoCon Europe 2023, I did [accessibility testing on the website’s homepage](https://thib.me/djangocon-europe-2023-accessibility-review-of-djangoproject-com). Here is a recap of issues I spotted, split into the categories I normally use, with links to my testing video’s timestamps, WCAG success criteria, and existing GitHub issues where appropriate.

Note though I only tested the homepage, there were very few content-related issues – I believe the vast majority of issues listed here are present site-wide.

### Blockers

> Blocker: Critical issue. Certain users cannot perceive or operate this, with no workaround available.

- The "hamburger" menu toggle is impossible to use with a keyboard. This makes it impossible for keyboard and screen reader users to access menu items and navigate the site in this way. ([Menu button is not a button #1234](https://github.com/django/djangoproject.com/issues/1234), [18:18](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1098s), [SC 2.1.1 Keyboard](https://www.w3.org/TR/WCAG22/#keyboard)).
- The page’s second column content is hidden in print stylesheets. This makes it impossible for users of magnifiers to access this content ([22:00](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1320s)).

### Major

> Major: Certain users are very likely to have difficulty perceiving or operating this.

- Heading level 1 should be used for the title of the page / visually most prominent heading, ideally not for invisible content ("Django" on homepage). Ideally each page would only have a single h1. ([07:04](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=424s), [SC 1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG22/#info-and-relationships)).

### Confusing

> Confusing: Understanding or operating this may be difficult for some users.

- Focus style almost invisible on "Get started with Django" CTA ([02:30](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=150s), [SC 1.4.11 Non-Text Contrast](https://www.w3.org/TR/WCAG22/#non-text-contrast)).
- Links should be identifiable with other means than color ([05:30](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=330s), [SC 1.4.1 Use of Color](https://www.w3.org/TR/WCAG22/#use-of-color)).
- All page content should be in a landmark ([09:39](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=579s), [SC 1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG22/#info-and-relationships)).
- A lot of text has color contrast issues making it hard to read ([#974](https://github.com/django/djangoproject.com/issues/974), [13:08](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=788s), [SC 1.4.3 Contrast (Minimum)](https://www.w3.org/TR/WCAG22/#contrast-minimum)).
- Scrolling should remain vertical only (no overflows) when using 400% page zoom ([16:48](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1008s), [SC 1.4.10 Reflow](https://www.w3.org/TR/WCAG22/#reflow)).
- Website logo disappears in forced colors mode with a light theme (white on white) ([19:14](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1154s), [SC 1.4.3 Contrast (Minimum)](https://www.w3.org/TR/WCAG22/#contrast-minimum)).

### Minor

> Minor: Degraded user experience for some users. This slows down use of the application.

- Missing skip link ([00:42](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=42s), [SC 2.4.1 Bypass Blocks](https://www.w3.org/TR/WCAG22/#bypass-blocks)).
- Custom "dotted outline" focus styles: not visible enough and broken in main navigation ([01:02](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=62s), [SC 2.4.13 Focus Appearance](https://www.w3.org/TR/WCAG22/#focus-appearance)).
- The two-column layout as implemented means reaching the second column content with the keyboard requires going through the entire first column ([02:58](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=178s), see also [#494](https://github.com/django/djangoproject.com/issues/494)).
- The sidebar heading hierarchy makes it hard to understand the content structure ([07:04](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=424s)).
- Keep uppercase text to a minimum. Can be problematic to read for some people, shouty, leads to pronounciation issues with screen readers ([15:40](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=940s)).
- Avoid interactive elements going right up to the edge of the screen on smaller viewports ("Get started with Django" button), as it makes it hard for touch users to touch-scroll past those elements ([17:30](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1050s)).
- For forced colors mode compatibility, show the regions of the page with borders since backgrounds are removed ([19:40](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1180s)).
- For forced colors mode compatibility, show the outline of buttons with borders since backgrounds are removed ([19:40](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1180s)).
- Form fields should remain visible in print stylesheets so the page content is understandable ([22:20](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1340s)).

### Best practices

> Best practices: These are not accessibility issues, but are recommended for ensuring the best user experience.

- The theme toggle’s icon should resize according to browser font size changes ([15:20](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=920s)).
- Footer logos should also ideally resize according to browser font size changes ([15:20](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=920s)).
- Hide the theme toggle in print styles ([21:05](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1265s)).
- Have link URLs visible in print styles ([21:05](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1265s)).
- Hide footer content in print styles ([21:05](https://www.youtube.com/watch?v=dAiMPTOMh1U&t=1265s)).

**Comment by thibaudcolas** [comment]
cc @tut-tuuut – if you have the chance it’d be great to hear of the above which ones we should prioritize. And if we could add things specific to the [Contact us page](https://www.djangoproject.com/contact/foundation/)

**Comment by tut-tuuut** [comment]
I created a PR for the contact us page : #2050

**Comment by tut-tuuut** [comment]
In A11y I like "quick wins" which help a lot of people. :)

Quick wins I could see in the short timespan I had:

- add a skip link to <main> element; the link can be visible only when focused. (Helps screenreader users but also keyboard navigators!)
- review of contrasts especially in header and footer which are everywhere.
- provide information not only relying on color and forms, especially in main navigation (see below).

---

### About the blocking topics mentioned above:

I didn't know a thing about print stylesheets and magnifiers tbh. I cannot confirm or deny the priority of this issue.

### About "confusing":

Heading levels: they look good to me, has it been fixed since the opening of this ticket?

Links should be identifiable with other means than color: true. They looked good to me though (underlining was added since this issue I guess?)

Focus indicators looked good to me too.

Generally, no information should be provided using only color or form (RGAA 10.10, WCAG 1.3.3 1.4.1). I've seen one main issue with this criteria in the header: active link is shown only using a different color.

Ideas to help colorblind ppl: add an underline or decoration of some sort
For screenreader users:
- with aria-current
- with a visually-hidden element
